File: BAEKJOON/src/silver/iv/java/BOJ_9012_Parentheses.java
package silver.iv.java;

import java.io.*;
import java.util.ArrayDeque;
import java.util.Deque;

// 괄호
public class BOJ_9012_Parentheses {

    private static boolean isVPS(String ps) {

        Deque<Character> stack = new ArrayDeque<>();
        for (int i = 0; i < ps.length(); i++) {
            char c = ps.charAt(i);
            if (c == '(') {
                stack.push(c);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }
                stack.pop();
            }
        }
        return stack.isEmpty();
    }

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        final int T = Integer.parseInt(br.readLine());
        for (int i = 0; i < T; i++) {
            bw.write(isVPS(br.readLine()) ? "YES" : "NO");
            bw.newLine();
        }
        br.close();
        bw.flush();
        bw.close();
    }
}

File: BAEKJOON/src/silver/iv/py/BOJ_9012_Parentheses.py
# 괄호
import sys


class Solution:
    @staticmethod
    def is_vps(ps):
        stack = []
        for c in ps:
            if c == '(':
                stack.append(c)
            else:
                if not stack:
                    return False
                stack.pop()
        return not stack


def main():
    for _ in range(int(sys.stdin.readline().strip())):
        print('YES' if Solution().is_vps(sys.stdin.readline().strip()) else 'NO')


if __name__ == '__main__':
    main()
